cupyx/scipy/sparse/compressed.py

```diff
@@ -558,10 +582,71 @@ def _get_major_slice(self, major):
         return self.__class__(
             (data, indices, indptr), shape=shape, dtype=self.dtype, copy=False)
 
-    @property
-    def has_canonical_format(self):
+    def __get_has_canonical_format(self):
+        """Determine whether the matrix has sorted indices and no duplicates
+
+        Returns
+            - True: if the above applies
+            - False: otherwise
+
+        has_canonical_format implies has_sorted_indices, so if the latter flag
+        is False, so will the former be; if the former is found True, the
+        latter flag is also set.
+
+        Modified from the SciPy counterpart.
+        """
+
+        # In CuPy the implemented conversions do not exactly match those of
+        # SciPy's, so it's hard to put this exactly as where it is in SciPy,
+        # but this should do the job.
+        if self.data.size == 0:
+            self._has_canonical_format = True
+        # check to see if result was cached
+        elif not getattr(self, '_has_sorted_indices', True):
+            # not sorted => not canonical
+            self._has_canonical_format = False
+        elif not hasattr(self, '_has_canonical_format'):
+            is_canonical = self._has_canonical_format_kern(
+                self.indptr, self.indices, size=self.indptr.size-1)
+            self._has_canonical_format = bool(is_canonical.all())
```

Comment:
This line synchronizes device to host. Please let me make sure that you think this sync is acceptable?

Reply:
Yes, that's right. It's necessary because internally we need this flag on the host to do control flow. But, the cost is most likely one-time thing (the first time this flag is encountered).

Reply:
Thanks. I think it's fine, then would you document about the synchronization as https://github.com/cupy/cupy/pull/2798?

Reply:
Ah, yes, thanks for reminder!

Reply:
@takagi I noticed that even though these attributes have docstrings, in actual usages like creating a `csr_matrix` and then calling `help(b.has_canonical_format`, for example, SciPy does not display it properly. I think it's because when subclassing the `_cs_matrix` (in SciPy's case), the docstrings should also be explicitly copied as follows:
```python
has_canonical_format.__doc__ = _cs_matrix.has_canonical_format.__doc__`
```
but I suspect they didn't do it because executing this line might trigger the actual computation. Should I simply document this in `docs/source/reference/sparse.rst ` instead?

Reply:
I think for the purpose of getting it documented on the website, Sphinx has done the job correctly. So I modified the getter docstrings. Let's see if it's OK.

Reply:
Yup: https://cupy--3509.org.readthedocs.build/en/3509/reference/generated/cupyx.scipy.sparse.csr_matrix.html#cupyx.scipy.sparse.csr_matrix.has_canonical_format

Reply:
Thanks! It's completely okay.

Reply:
Can I ask you to fix the docstrings of `sum_duplicates()` and `sort(ed)_indices()` methods, which use the properties and might also synchronize the device after this PR is merged?

Reply:
Definitely!